### PR TITLE
RangeControl: Convert stories to TypeScript

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
Depends on:
- https://github.com/WordPress/gutenberg/pull/40535

Related:
- https://github.com/WordPress/gutenberg/issues/40507
- https://github.com/WordPress/gutenberg/pull/41445

## What?

Converts `RangeControl` storybook examples to TypeScript and updates them to match the latest approach to component stories.

This is based off #40535 in order to separate discussion on storybook examples from the main effort of typing the `RangeControl`.

## Why?

It's another step closer to typing the components package and improving the Storybook as living documentation.

## How?

- Converted to TS
- Reworked stories

## Testing Instructions

1. Explore the component functionality and each example via the Storybook: `npm run storybook:dev`

### CODE REVIEW

**Review by ciampo — COMMENTED** [review]
Thank you for working on this, Aaron! It's already looking great :)

I'll leave a few observations here, mostly related to the specific types and controls displayed in the auto-generated docs:

- the type that is being shown for `afterIcon` and `beforeIcon` is `IconType<unknown>`, which is not very useful documentation — although I'm not sure if we can do much about it, since the type is defined in the `Icon` component (this is a known limitation of the docgen employed by Storybook). Maybe we can add some text to the prop description text, referencing the `Icon` component for more info ?
- the control being used for the `as` prop is a dropdown containing all the possible HTML elements. This is interesting, because in other components the docgen was not able to determine this list, and was simply using the "Object" type — because of that, we previously had to [override the control type to be a simple string of text for the `as` prop](https://github.com/WordPress/gutenberg/blob/82c0ef49cd2fbf37299598f70ab805e65c4091bf/packages/components/src/heading/stories/index.tsx#L16). It would be interesting to understand the reason for this different behaviour
- as a side note, do we think that it makes sense for `RangeControl` to be polymorphic (i.e. to accept the `as` prop)? Any other value than `input` kind of breaks the component.
- the color-related props' type (`color`, `railColor`, `trackColor`) is displayed as `Color`. Maybe we could add some text in the description to mention that the prop accepts any CSS color string?
- There's no need to wrap the default `COLORS.ui.theme` value with backticks in the props JSDocs
- The fact that the docgen can't "expand" compound types makes it so that the `RangeStep`, `RangeMarks` and `RenderTooltipContentCallback` types are displayed in the docs, giving little information about what the actual type for the given prop is. I know this is not very elegant, but we could look into removing the type alias and leaving the expanded versions in the prop types?
- the control for the `type` prop is currently an "Object". Since that prop's value can only be `"stepper"` or `undefined`, we should switch to a more fitting control (e.g. checkbox / radio?)

File: packages/components/src/range-control/stories/index.tsx [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+	initialPosition: 50,
+	label: 'Opacity',
+	max: 100,
+	min: 0,
+};
+
+/**
+ * Setting the `step` prop to `"any"` will allow users to select non-integer
+ * values. This also overrides both `withInputField` and `showTooltip` props to
+ * `false`.
```
Comment (mirka) [inline packages/components/src/range-control/stories/index.tsx:87]: I think we have a regression in the base branch of this PR, because this is now showing the tooltip!

Story with any step is showing tooltip

File: packages/components/src/range-control/stories/index.tsx [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+		initialPosition: { control: { type: 'number' } },
+		marks: { control: { type: 'object' } },
+		min: { control: { type: 'number' } },
+		max: { control: { type: 'number' } },
+		onChange: { action: 'onChange', control: { type: null } },
+		railColor: { control: { type: 'color' } },
+		showTooltip: { control: { type: 'boolean' } },
+		step: { control: { type: 'number' } },
+		trackColor: { control: { type: 'color' } },
+		withInputField: { control: { type: 'boolean' } },
```
Comment (mirka) [inline packages/components/src/range-control/stories/index.tsx]: A bunch of these can be removed, now that we have type data:

```suggestion
		beforeIcon: {
			control: { type: 'select' },
			options: Object.keys( ICONS ),
			mapping: ICONS,
		},
		color: { control: { type: 'color' } },
		help: { control: { type: 'text' } },
		marks: { control: { type: 'object' } },
		onChange: { action: 'onChange' },
		railColor: { control: { type: 'color' } },
		step: { control: { type: 'number' } },
		trackColor: { control: { type: 'color' } },
```

On the other hand, I guess we should disable the controls for these new callback props: `onFocus`, `onMouseLeave`, `onMouseMove`, `renderTooltipContent`. (Storybook disables controls on `on*` callback props by default, but in these cases I think it doesn't understand the types as function types?)

Also, we should disable the `value` control now that our stories are controlled.

**Review by mirka — APPROVED** [review]
Very nice improvements!

**Comment by mirka** [comment]
> * as a side note, do we think that it makes sense for `RangeControl` to be polymorphic (i.e. to accept the `as` prop)? Any other value than `input` kind of breaks the component.

Good point! I'd want to remove the polymorphism, but this is already a stable component 😭 

> * the color-related props' type (`color`, `railColor`, `trackColor`) is displayed as `Color`. Maybe we could add some text in the description to mention that the prop accepts any CSS color string?

Personally I'm fine with the current description, especially since the control right next to it is a standard color picker. If we are going to rewrite the descriptions though, a sentence structure like "CSS color string for ____" would probably work without being too wordy.

**Comment by ciampo** [comment]
> Good point! I'd want to remove the polymorphism, but this is already a stable component 😭

I think that the `as` prop is being added by the `WordPressComponentProps` utility type in https://github.com/WordPress/gutenberg/pull/40535 — if that's the case, then we can safely remove it, since that hasn't been merged in `trunk` yet

File: packages/components/src/range-control/stories/index.tsx (hunk 1 above)
Comment (aaronrobertshaw) [reply packages/components/src/range-control/stories/index.tsx:87]: Good catch. Thank you! I've fixed this in the base branch.

File: packages/components/src/range-control/stories/index.tsx (hunk 2 above)
Comment (aaronrobertshaw) [reply packages/components/src/range-control/stories/index.tsx]: Thanks, I've cleaned up the controls here. ( https://github.com/WordPress/gutenberg/pull/41444/commits/9631a22b8d87b28f5285cf8153af2c687147d608 & https://github.com/WordPress/gutenberg/pull/41444/commits/7e06827ab81a020f81106ced4fd8283048a32610)

**Comment by aaronrobertshaw** [comment]
> I think that the as prop is being added by the WordPressComponentProps utility type in https://github.com/WordPress/gutenberg/pull/40535 — if that's the case, then we can safely remove it, since that hasn't been merged in trunk yet

I've omitted the `as` prop from `WordPressComponentProps` in the base branch. Let me know if that suffices.

**Review by ciampo — APPROVED** [review]
LGTM 🎉

**Comment by aaronrobertshaw** [comment]
Thanks for the reviews @ciampo, I've had some unexpected AFK but will merge these next week when I'm back on deck.

File: packages/components/src/range-control/stories/index.tsx [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+			options: Object.keys( ICONS ),
+			mapping: ICONS,
+		},
+		beforeIcon: {
+			control: { type: 'select' },
+			options: Object.keys( ICONS ),
+			mapping: ICONS,
+		},
+		color: { control: { type: 'color' } },
+		help: { control: { type: 'text' } },
```
Comment (ciampo) [inline packages/components/src/range-control/stories/index.tsx:34]: Since dashicons are not loaded in Storybook, it's probably better to disable this control?

```suggestion
		help: { control: { type: 'text' } },
		icon: { control: { type: null } },
```

Comment (aaronrobertshaw) [reply packages/components/src/range-control/stories/index.tsx:34]: ✅ Done, thanks.
